src/rdb_protocol/terms/string.cc

```diff
@@ -125,6 +127,55 @@ class match_term_t : public op_term_t {
     virtual const char *name() const { return "match"; }
 };
 
+class format_term_t : public op_term_t {
+public:
+    format_term_t(compile_env_t *env, const raw_term_t &term)
+            : op_term_t(env, term, argspec_t(2)) {}
+private:
+    virtual scoped_ptr_t <val_t> eval_impl(scope_env_t *env, args_t *args, eval_flags_t) const {
+        const std::string templ = args->arg(env, 0)->as_str().to_std();
+        const datum_t datum = args->arg(env, 1)->as_datum();
+
+        std::map <datum_string_t, datum_t> params{};
+        for (size_t i = 0; i < datum.obj_size(); ++i) {
+            const auto datum_pair{datum.get_pair(i)};
+            rcheck_target(&datum_pair.second, datum_pair.second.get_type() == datum_t::R_STR,
+                          base_exc_t::LOGIC,
+                          strprintf("Cannot call `%s` on objects of type `%s`.",
+                                    name(), datum_pair.second.get_type_name().c_str()));
+
+            params.insert(datum_pair);
+        }
```

Comment:
I don't see why we should copy out all the objects from the datum.  Instead, just use get_field on the datum.

Also, if we were to mandate all formatted fields be strings (or some other requirement), why should we mandate all the properties be strings and not just the ones we're formatting?  Generally, I think people might pass in documents that are not specially trimmed for the format command.

Reply:
Oh! I missed the get_field. Also, fair point to validate only necessary params.

If we would coerce the params to strings how would you do that to not loose/add extra precision for doubles?